drivers/crypto/crypto_nxp_s32_hse.c

```diff
@@ -0,0 +1,618 @@
+		    (hseSrvDescriptor_t *)&session->crypto_serv_desc) != HSE_SRV_RSP_OK) {
+		k_mutex_unlock(&session->crypto_lock);
+		return -EIO;
+	}
+
+	memcpy(pkt->out_buf, session->out_buff,
+	       *(uint32_t *)(session->crypto_serv_desc.hseSrv.hashReq.pHashLength));
+
+	k_mutex_unlock(&session->crypto_lock);
+
+	return 0;
+}
+
+static int crypto_nxp_s32_hse_hash_begin_session(const struct device *dev, struct hash_ctx *ctx,
+						 enum hash_algo algo)
+{
+	struct crypto_nxp_s32_hse_session *session;
+	static uint32_t out_len;
+	hseHashSrv_t *hash_serv;
+	struct k_mutex crypto_lock;
+
+	if (ctx->flags & ~(CRYPTO_NXP_S32_HSE_HASH_CAPS)) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_init(&crypto_lock);
+	k_mutex_lock(&crypto_lock, K_FOREVER);
+	session = crypto_nxp_s32_hse_get_session(dev);
+	k_mutex_unlock(&crypto_lock);
+
+	if (session == NULL) {
+		return -ENOSPC;
+	}
+
+	session->req_type.eReqType = HSE_IP_REQTYPE_SYNC;
+	session->req_type.u32Timeout = CRYPTO_NXP_S32_HSE_TIMEOUT;
+
+	hash_serv = &(session->crypto_serv_desc.hseSrv.hashReq);
+
+	switch (algo) {
+	case CRYPTO_HASH_ALGO_SHA224: {
+		hash_serv->hashAlgo = HSE_HASH_ALGO_SHA2_224;
+		out_len = HSE_BITS_TO_BYTES(224);
+		break;
+	}
+	case CRYPTO_HASH_ALGO_SHA256: {
+		hash_serv->hashAlgo = HSE_HASH_ALGO_SHA2_256;
+		out_len = HSE_BITS_TO_BYTES(256);
+		break;
+	}
+	case CRYPTO_HASH_ALGO_SHA384: {
+		hash_serv->hashAlgo = HSE_HASH_ALGO_SHA2_384;
+		out_len = HSE_BITS_TO_BYTES(384);
+		break;
+	}
+	case CRYPTO_HASH_ALGO_SHA512: {
+		hash_serv->hashAlgo = HSE_HASH_ALGO_SHA2_512;
+		out_len = HSE_BITS_TO_BYTES(512);
+		break;
+	}
+	default: {
+		return -ENOTSUP;
+	}
+	}
+
+	session->crypto_serv_desc.srvId = HSE_SRV_ID_HASH;
+	hash_serv->accessMode = HSE_ACCESS_MODE_ONE_PASS;
+	hash_serv->pHashLength = HSE_PTR_TO_HOST_ADDR(&out_len);
+	hash_serv->pHash = HSE_PTR_TO_HOST_ADDR(session->out_buff);
+
+	ctx->drv_sessn_state = session;
+	ctx->hash_hndlr = crypto_nxp_s32_hse_sha;
+	ctx->device = dev;
+
+	return 0;
+}
+
+static int crypto_nxp_s32_hse_hash_free_session(const struct device *dev, struct hash_ctx *ctx)
+{
+	struct crypto_nxp_s32_hse_session *session;
+	struct crypto_nxp_s32_hse_data *data = dev->data;
+
+	session = ctx->drv_sessn_state;
+	k_mutex_lock(&session->crypto_lock, K_FOREVER);
+	memset(&session->req_type, 0, sizeof(Hse_Ip_ReqType));
+	memset(&session->crypto_serv_desc, 0, sizeof(hseSrvDescriptor_t));
+	Hse_Ip_ReleaseChannel(data->mu_instance, session->channel);
+	free_session(session);
+	k_mutex_unlock(&session->crypto_lock);
+
+	return 0;
+}
+
+static int crypto_nxp_s32_hse_query_caps(const struct device *dev)
+{
+	return CRYPTO_NXP_S32_HSE_HASH_CAPS | CRYPTO_NXP_S32_HSE_CIPHER_CAPS;
+}
+
+static int crypto_nxp_s32_hse_init(const struct device *dev)
+{
+	struct crypto_nxp_s32_hse_data *data = dev->data;
+	struct crypto_nxp_s32_hse_session *session;
+
+	if (Hse_Ip_Init(data->mu_instance, &data->mu_state) != HSE_IP_STATUS_SUCCESS) {
```

Comment:
It's not enough to just call this initialiation but also the MUs that are not active by default must be activated through a request to HSE. As it is right now implemented, if the user tries to use the Zephyr Crypto API on a HSE-MU device that is not active, it will fail. In general, the device probe function should ensure that all needed resources are allocated and ready in order for the driver to operate after exiting this function.